### PR TITLE
Added example on train_step sharing across multiple models

### PR DESCRIPTION
**Pull request by nkovela1**
This is an addition to Quick Keras recipes in response to requests to document train step sharing across multiple models using the Trainer pattern. (b/226021975)

### CODE REVIEW

File: examples/keras_recipes/trainer_pattern.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+YourModel = tf.keras.Model(func_input, func_output)
+
+"""
+## Create Trainer class objects from the models
+"""
+trainer_1 = MyTrainer(MyModel)
+trainer_2 = MyTrainer(YourModel)
+
+"""
+## Compile and fit the models to the MNIST dataset
```
Comment (fchollet) [inline examples/keras_recipes/trainer_pattern.py:69]: Please follow-up with a more advanced example of the Trainer pattern, such as a Trainer for a model that has two models attached to it. Maybe we could draw from this GAN example -- or just link to it https://keras.io/guides/customizing_what_happens_in_fit/#wrapping-up-an-endtoend-gan-example

File: examples/keras_recipes/trainer_pattern.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+(x_train, y_train),(x_test, y_test) = mnist.load_data()
+x_train, x_test = x_train / 255.0, x_test / 255.0
+
+
+"""
+## Sharing custom training step across multiple models
+
+A custom training step can be created by overriding the train_step method of a Model subclass.
+""" 
+class MyTrainer(tf.keras.Model):
```
Comment (fchollet) [inline examples/keras_recipes/trainer_pattern.py:26]: Import `keras` then use `keras.something`

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: examples/keras_recipes/trainer_pattern.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        with tf.GradientTape() as tape:
+            y_pred = self.model(x, training=True)
+            loss = self.compiled_loss(y, y_pred)
+        self.optimizer.minimize(loss, self.trainable_variables, tape=tape)
+        return self.compute_metrics(x, y, y_pred, None)
+
+"""
+## Define multiple models to share the custom training step
+
+Let's define two different models that can share our Trainer class and its custom train_step
```
[the remark this answers is not in the document]
Comment (nkovela1) [reply examples/keras_recipes/trainer_pattern.py:67]: Added more explanations.

**Review by fchollet — COMMENTED** [review]
Thanks for the updates!

File: examples/keras_recipes/trainer_pattern.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+class MyTrainer(tf.keras.Model):
+    def __init__(self, model):
+        super().__init__()
+        self.model = model
+
+    def train_step(self, data):
+        x, y = data
+        with tf.GradientTape() as tape:
+            y_pred = self.model(x, training=True)
+            loss = self.compiled_loss(y, y_pred)
```
[the remark this answers is not in the document]
Comment (nkovela1) [reply examples/keras_recipes/trainer_pattern.py]: Fixed.

File: examples/keras_recipes/trainer_pattern.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        self.optimizer.minimize(loss, self.trainable_variables, tape=tape)
```
[the remark this answers is not in the document]
Comment (nkovela1) [reply examples/keras_recipes/trainer_pattern.py]: Fixed.

File: examples/keras_recipes/trainer_pattern.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+"""
+## Setup
+"""
+
+import tensorflow as tf
+
+# Load MNIST dataset and normalize
+mnist = tf.keras.datasets.mnist
+(x_train, y_train),(x_test, y_test) = mnist.load_data()
```
[the remark this answers is not in the document]
Comment (nkovela1) [reply examples/keras_recipes/trainer_pattern.py]: Fixed in latest commit.

File: examples/keras_recipes/trainer_pattern.py (hunk 1 above)
Comment (nkovela1) [reply examples/keras_recipes/trainer_pattern.py:104]: Link added in Introduction.

File: examples/keras_recipes/trainer_pattern.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
+"""
+Title: Trainer pattern
+Author: [nkovela1](https://nkovela1.github.io/)
+Date created: 2022/09/19
+Last modified: 2022/09/19
+Description: Demonstration of sharing train_step across multiple Keras Models.
```
[the remark this answers is not in the document]
Comment (nkovela1) [reply examples/keras_recipes/trainer_pattern.py]: Fixed.

File: examples/keras_recipes/trainer_pattern.py (hunk 2 above)
Comment (nkovela1) [reply examples/keras_recipes/trainer_pattern.py]: Fixed.

File: examples/keras_recipes/trainer_pattern.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+])
+
+# A model defined using Functional API
+func_input = keras.Input(shape=(28,28,1))
+x = keras.layers.Flatten(input_shape=(28,28))(func_input)
+x = keras.layers.Dense(512, activation='relu')(x)
+x = keras.layers.Dropout(0.4)(x)
+func_output = keras.layers.Dense(10, activation='softmax')(x)
+
+YourModel = keras.Model(func_input, func_output)
```
[the remark this answers is not in the document]
Comment (nkovela1) [reply examples/keras_recipes/trainer_pattern.py]: Renamed to model_b. Thanks!

File: examples/keras_recipes/trainer_pattern.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+	return x
+
+"""
+## Define multiple models to share the custom training step
+
+Let's define two different models that can share our Trainer class and its custom train_step
+"""
+
+# A model defined using Sequential API
+MyModel = keras.models.Sequential([
```
[the remark this answers is not in the document]
Comment (nkovela1) [reply examples/keras_recipes/trainer_pattern.py]: Renamed to model_a. Thanks!

File: examples/keras_recipes/trainer_pattern.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+Title: Trainer pattern
+Author: [nkovela1](https://nkovela1.github.io/)
+Date created: 2022/09/19
+Last modified: 2022/09/23
+Description: Guide on how to share a custom training step across multiple Keras models.
+"""
+"""
+## Introduction
+
+This example shows how to create a custom training step using the Trainer pattern, which can then be shared across multiple Keras models. This pattern overrides the train_step method of the keras.Model class, allowing for the addition of custom loss functions and metrics.
```
[the remark this answers is not in the document]
Comment (nkovela1) [reply examples/keras_recipes/trainer_pattern.py]: Fixed.

**Review by fchollet — COMMENTED** [review]
Thanks for the updates! I pushed some copyedits. Please pull them, then add the generated files.

**Pull request merged**